### PR TITLE
bugfix: StatePrep.state_vector works with backprop

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
First part needed in the fix for #4964. The `StatePrep.state_vector` function should be differentiable, but isn't currently. The second part appears to be because `adjoint_jacobian` explicitly doesn't work for state-prep operations.

**Description of the Change:**
Re-write `StatePrep.state_vector` so it is backprop-friendly

**Benefits:**
A state-prep operator acting on a subset of all circuit wires won't break backprop with torch

### CODE REVIEW

File: doc/releases/changelog-0.34.0.md [hunk 1]
```diff
@@ -719,6 +719,8 @@
 * The jax-jit interface can now be used with float32 mode.
   [(#4990)](https://github.com/PennyLaneAI/pennylane/pull/4990)
 
+* `StatePrep` no longer raises unexpected errors during forward calls with `torch.nn`.
```
Comment (timmysilv) [inline doc/releases/changelog-0.34.0.md:722]:
```suggestion
* `StatePrep` no longer raises unexpected errors during forward calls with `torch.nn`.
  [(#4996)](https://github.com/PennyLaneAI/pennylane/pull/4996)
```

File: doc/releases/changelog-0.34.0.md [hunk 2]
```diff
@@ -719,6 +719,9 @@
...
```
Comment (albi3ro) [inline doc/releases/changelog-0.34.0.md:722]:
```suggestion
* `StatePrep` operations expanded onto more operations are now compatible with backrpop.
```

File: pennylane/ops/qubit/state_preparation.py [hunk 3]
```diff
@@ -227,7 +227,7 @@ def state_vector(self, wire_order=None):
             ket_shape = [self.batch_size] + ket_shape
 
         ket = np.zeros(ket_shape, dtype=np.complex128)
-        ket[indices] = op_vector
+        ket[indices] = math.unwrap(op_vector)
```
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py:230]: Wouldn't this break differentiation?

Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py]: not anymore! totally re-vamped the function.

File: tests/ops/qubit/test_state_prep.py [hunk 4]
```diff
@@ -191,6 +191,56 @@ def test_StatePrep_wrong_param_size_fails(self):
         with pytest.raises(ValueError, match="State vector must have shape"):
             _ = qml.StatePrep([0, 1], wires=[0, 1])
 
+    @pytest.mark.torch
```
Comment (albi3ro) [inline tests/ops/qubit/test_state_prep.py:194]: I'm not sure we need the whole qnn test.  The bug itself can just be checked by making sure it's backprop compatible. Then again, can't hurt.

**Pull request closed without merging**